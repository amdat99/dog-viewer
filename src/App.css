.App {
  text-align: center;
  
}

button{
  border-color: #0096aa;
  background: #0096aa;
  color: #ffffff;
  cursor: pointer;
  border-radius:5px;
  border: transparent;
  width: 100px;
  height: 40px; 
}

button:hover{
  background: #ffffff;
  color: #0096aa;
  margin-left: 20px;
  border:1px solid #0096aa;
}

select{
  outline: none;
  border: 1px solid rgba(0, 0, 0, 0);
  box-shadow: 2px 2px 2px 2px hsl(218, 60%, 92%);
  height: 25px;

}
.container{
  display:flex;

  margin-top:20px;
  margin-left: 20px;
}

.form-group{
  display:flex;
  flex-direction: row;
  justify-content: flex-start;

}

.selector{
  display:flex;
  flex-direction: column;
  justify-content: space-evenly;
  margin-left: 10px;
}

.form-control{
  margin-top: 5px;
  position: relative;
 
}


#error-border{

  border: 1px solid #ff0000;
}