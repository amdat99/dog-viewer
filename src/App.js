import React, {useState, useEffect} from 'react';
import { LazyLoadImage } from 'react-lazy-load-image-component';


import Selectors from './selectors/Selectors';
import {fetchData} from './utils'

import './App.css';

function App() {
  
  const [breeds, setBreeds] = useState(null);
  const [currentBreed , setCurrentBreed] = useState('')
  const [currentSubBreedList, setCurrentSubBreedList] = useState(null);

  const [currentSubBreed , setCurrentSubBreed] = useState('')
  const [imagesData, setImagesData] = useState(['https://images.dog.ceo/breeds/hound-afghan/n02088094_10822.jpg'])
  const [numOfImgs, setNumOfImgs] = useState(null)
  const [currentImgNum,setCurrentImgNum] = useState(1)



  useEffect(() => {
    fetchData('https://dog.ceo/api/breeds/list/all').then(data =>{
    setBreeds(Object.keys(data.message))
  })
  }, []);
  
  
  useEffect(() => {
    if (breeds){
      setCurrentBreed(breeds[0])
    }
  },[setBreeds,breeds])

  const fetchImages = (e) =>{

  if(!currentBreed || !currentImgNum){
  return alert('Please select a breed and number of images')
  }
  e.preventDefault()


  if(currentBreed){
    if(currentSubBreed){ 
   
      fetchData(`https://dog.ceo/api/breed/${currentBreed}/${currentSubBreed}/images/random/${currentImgNum}`).then(data => {
        if(data){
        setImagesData(data.message)
        }

            })
    } else{
      fetchData(`https://dog.ceo/api/breed/${currentBreed}/images/random/${currentImgNum}`).then(data => {
        if (data){
       setImagesData(data.message)
      }
      })}
  }
}

const getImageLength = () =>{

  if(currentSubBreed){ 
   
    fetchData(`https://dog.ceo/api/breed/${currentBreed}/${currentSubBreed}/images`).then(data => {
    setNumOfImgs(data.message.length)
  })
  } else{
  fetchData(`https://dog.ceo/api/breed/${currentBreed}/images`).then(data => {
    setNumOfImgs(data.message.length)
  })}
  
}
return (
 
    <div className="container">
        
        <form className="form-group" onSubmit={fetchImages}>
         
          <Selectors currentBreed={currentBreed} currentSubBreedList={currentSubBreedList} breeds={breeds} setCurrentImgNum = {setCurrentImgNum} 
          setCurrentSubBreedList={setCurrentSubBreedList}  setCurrentBreed={setCurrentBreed} setCurrentSubBreed={setCurrentSubBreed} 
         currentSubBreed={currentSubBreed} numOfImgs={numOfImgs} getImageLength={getImageLength} currentImgNum={currentImgNum} />
        </form> 
    
      <div className="images-viewer">
          {imagesData?
              imagesData.map((image, i) => {
              return <LazyLoadImage className="image-toggle" src={image} key={i} width={200} height={200}/> 
          })
         :null}
      </div>
         
     </div>
    
  );
}

export default App;
