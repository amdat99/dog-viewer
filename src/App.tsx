import React, { useState, useEffect } from "react";
import { LazyLoadImage } from "react-lazy-load-image-component";

import Selectors from "./selectors/Selectors";
import Modal from "./modal/Modal";

import { fetchData } from "./utils";

import "./App.css";

function App() {
  const [breeds, setBreeds] = useState<string[]>([]);
  const [currentBreed, setCurrentBreed] = useState("");
  const [currentSubBreedList, setCurrentSubBreedList] = useState([]);
  const [currentSubBreed, setCurrentSubBreed] = useState("");
  const [imagesData, setImagesData] = useState([]);
  const [numOfImgs, setNumOfImgs] = useState(1);
  const [currentImgNum, setCurrentImgNum] = useState(1);
  const [toggleModal, setToggleModal] = useState(false);
  const [modalImage, setModalImage] = useState("");
  // const [showDownloadIcon, setShowDownloadIcon] = useState(false)
  const [scrollX, setScrollX] = useState(0);

  useEffect(() => {
    fetchData("https://dog.ceo/api/breeds/list/all").then((data) => {
      setBreeds(Object.keys(data.message));
    });
  }, []);

  useEffect(() => {
    if (breeds) {
      setCurrentBreed(breeds[0]);
      console.log(currentBreed);
    }
  }, [setBreeds, breeds, currentBreed]);

  const fetchImages = (e: React.FormEvent<HTMLInputElement>) => {
    e.preventDefault();
    if (!currentBreed || !currentImgNum) {
      return alert("Please select a breed and number of images");
    }

    if (currentBreed) {
      if (currentSubBreed) {
        console.log(currentSubBreed, currentBreed, currentImgNum);
        fetchData(
          `https://dog.ceo/api/breed/${currentBreed}/${currentSubBreed}/images/random/${currentImgNum}`
        ).then((data) => {
          if (data) {
            setImagesData(data.message);
          }
        });
      } else {
        fetchData(
          `https://dog.ceo/api/breed/${currentBreed}/images/random/${currentImgNum}`
        ).then((data) => {
          if (data) {
            setImagesData(data.message);
          }
        });
      }
    }
  };

  const getImageLength = () => {
    if (currentSubBreed) {
      fetchData(
        `https://dog.ceo/api/breed/${currentBreed}/${currentSubBreed}/images`
      ).then((data) => {
        setNumOfImgs(data.message.length);
      });
    } else {
      fetchData(`https://dog.ceo/api/breed/${currentBreed}/images`).then(
        (data) => {
          setNumOfImgs(data.message.length);
        }
      );
    }
  };

  const initialiseModalImage = async (image: string) => {
    await setModalImage(image);
    await setToggleModal(true);
    setScrollX(document.documentElement.scrollTop);
    window.scroll({
      top: 0,
      left: 0,
      behavior: "smooth",
    });
  };

  const closeModal = () => {
    setToggleModal(false);
    window.scroll({
      top: scrollX,
      left: 0,
      behavior: "smooth",
    });
  };

  async function downloadImage(imageSrc: string) {
    const image = await fetch(imageSrc);
    const imageBlog = await image.blob();
    const imageURL = URL.createObjectURL(imageBlog);

    const link = document.createElement("a");
    link.href = imageURL;
    link.download = currentBreed + "-" + currentSubBreed;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
  }
  return (
    <>
      {toggleModal ? (
        <div onClick={closeModal} className={"background-cover"}></div>
      ) : null}

      <div className="container">
        <form
          className="form-group"
          // @ts-ignore */
          onSubmit={fetchImages}
        >
          <Selectors
            currentBreed={currentBreed}
            currentSubBreedList={currentSubBreedList}
            breeds={breeds}
            setCurrentImgNum={setCurrentImgNum}
            // @ts-ignore */
            setCurrentSubBreedList={setCurrentSubBreedList}
            setCurrentBreed={setCurrentBreed}
            setCurrentSubBreed={setCurrentSubBreed}
            currentSubBreed={currentSubBreed}
            numOfImgs={numOfImgs}
            getImageLength={getImageLength}
            currentImgNum={currentImgNum}
          />

          <button id="view-button" type="submit">
            View Images
          </button>
        </form>

        <div className="images-viewer">
          {imagesData
            ? imagesData.map((image: string, i: number) => {
                return (
                  <LazyLoadImage
                    // onMouseEnter={() => setShowDownloadIcon(true)}
                    // onMouseLeave={() => setShowDownloadIcon(false)}
                    className="image-toggle"
                    src={image}
                    key={i}
                    width={200}
                    height={200}
                    onClick={() => initialiseModalImage(image)}
                  />
                );
              })
            : null}
        </div>
        {toggleModal ? (
          <Modal>
            {/* <span style={{color: 'red',position: 'absolute' ,top:'-3%', right: '10%'}}>close</span> */}
            <img
              src={modalImage}
              alt="clicked-media"
              className="modal-image"
              onClick={closeModal}
            />
            <span
              style={{ textDecoration: "none", cursor: "pointer" }}
              onClick={() => downloadImage(modalImage)}
            >
              {" "}
              ðŸ’¾
            </span>
            <a
              data-pgaction-redirection="0"
              target="_blank"
              rel="noopener noreferrer"
              title={"dog image"}
              href={`https://www.facebook.com/sharer.php?u=${window.location.href}?imageurl=${modalImage}`}
            >
              Facebook
            </a>
          </Modal>
        ) : null}
      </div>
    </>
  );
}

export default App;
