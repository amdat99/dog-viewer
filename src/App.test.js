import { useEffect } from "react";
import { withHooks } from "jest-react-hooks-shallow";

import { shallow, render, mount } from "enzyme";
import { createSerializer } from "enzyme-to-json";
import toJson from "enzyme-to-json";

import App from "./App.js";

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

let mockFetchImages = jest.fn();
let mockFetchData = jest.fn();

let wrapper2;
let wrapper3;

const formEventMocked = { preventDefault: jest.fn() };
window.alert = jest.fn();

beforeEach(() => {
  const mockProps = {
    message: [
      "https://images.dog.ceo/breeds/ovcharka-caucasian/IMG_20190528_194200.jpg",
      "https://images.dog.ceo/breeds/ovcharka-caucasian/IMG_20190601_185700.jpg",
    ],

    fetchImages: mockFetchImages,
    fetchData: mockFetchData,
  };

  // wrapper = toJson(shallow(<App {...mockProps} />));
  wrapper2 = shallow(<App {...mockProps} />);
  wrapper3 = render(<App {...mockProps} />);
});

it("expect to render App component", () => {
  expect(wrapper2).toMatchSnapshot();

  console.log(wrapper2.debug());
});

it("expect alert to be called and img source undefined when fetchImages called ", () => {
  wrapper2.find("form").simulate("submit", formEventMocked);
  expect(wrapper3.find("image-toggle").prop("src")).toEqual(undefined);
  window.alert.mockClear();
});
