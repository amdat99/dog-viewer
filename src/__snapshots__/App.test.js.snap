// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render App component 1`] = `
<Fragment>
  <div
    className={null}
    onClick={null}
  />
  <div
    className="container"
  >
    <form
      className="form-group"
      onSubmit={[Function]}
    >
      <Selectors
        breeds={Array []}
        currentImgNum={1}
        currentSubBreed=""
        currentSubBreedList={null}
        getImageLength={[Function]}
        numOfImgs={null}
        setCurrentBreed={[Function]}
        setCurrentImgNum={[Function]}
        setCurrentSubBreed={[Function]}
        setCurrentSubBreedList={[Function]}
      />
      <button
        id="view-button"
        type="submit"
      >
        View Images
      </button>
    </form>
    <div
      className="images-viewer"
    />
  </div>
</Fragment>
`;
